File: elixir/sum-of-multiples/lib/sum_of_multiples.ex
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    0..limit - 1
    |> Stream.filter(&(multiple?(factors, &1)))
    |> Stream.take(limit)
    |> Enum.sum
  end

  defp multiple?(factors, 0), do: 0 in factors
  defp multiple?(factors, n) do
    factors
    |> Stream.filter(&(&1 > 0))
    |> Enum.any?(&(rem(n, &1) == 0))
  end
end

File: c/sum-of-multiples/sum_of_multiples.c
#include "sum_of_multiples.h"
#include <stdbool.h>

unsigned int sum(const unsigned int *factors, const size_t number_of_factors,
                 const unsigned int limit) {
  if (number_of_factors <= 0 || limit <= 1 || factors == NULL) {
    return 0;
  }
  bool multiples[limit];
  for (unsigned int i = 1; i < limit; ++i) {
    multiples[i] = false;
    for (size_t j = 0; j < number_of_factors; ++j) {
      if (factors[j] == 0) {
        continue;
      }
      if (i % factors[j] == 0) {
        multiples[i] = true;
      }
    }
  }
  unsigned int sum = 0;
  for (unsigned int i = 1; i < limit; ++i) {
    if (multiples[i]) {
      sum += i;
    }
  }
  return sum;
}
